### PR TITLE
Overhaul

### PR DESCRIPTION
**Pull request by Danilcus-W**
# About the pull request

- AI peoples now get qdeleted if there was no chosen appearance, e.g. cancelled
- AI preset now appraises inventory after loading for correct inventory logging
- AI now sets target when shot from afar, blasting local ~~cheesers~~ snipers off
- Pumping and bolt actions are now listed as before_fire() procs in weapon appraisals
- Introducing optimal range var, replacing maximum range at approaching action in attack loop, so we can start shooting before stepping in a desired position
- When target is prone, we consider it mortally wounded, and approach at minimal range of our weapon to finish it off faster (I guess this var was made to tell AI when to start backing off, but nothing stops us from using it in two places simultaneously in the future [its 6 A.M. and im too tired to code retreat action right now])
- Set register signal override to true as it's required when we can accidentally register two similar signals in our loops
- Changed in_chamber check to current_ammo check as many guns have no interactions with gun chamber, such as flamer
- AI Humans now face in direction of their fire
- Utility belts are now appropriately listed by AI inventory
- AI can use flamers
- AI can use RPG's

# Explain why it's good for the game

It's not, i saw CLF doing horrible things to civilians

# Testing Photographs and Procedure

<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

# Changelog

If i remember correctly, merging a changelog will fuck with the PR of this branch on PVE, so i'll leave it blank

### CODE REVIEW

File: code/modules/mob/living/carbon/human/ai/ai_equipment.dm [hunk 1]
```diff
@@ -16,10 +16,24 @@
...
+
 	new_human.equip_to_slot_or_del(new /obj/item/storage/belt/marine(new_human), WEAR_WAIST)
 	if(prob(50))
 		spawn_rebel_smg(new_human)
 	else
 		spawn_rebel_rifle(new_human)
 
 	new_human.equip_to_slot_or_del(new /obj/item/device/radio/headset/distress/CLF(new_human), WEAR_L_EAR)
+
+/datum/equipment_preset/clf/soldier/ai/load_preset(mob/living/carbon/human/new_human, randomise, count_participant, client/mob_client, show_job_gear)
```
Comment (Zonespace27) [inline code/modules/mob/living/carbon/human/ai/ai_equipment.dm:36]: i'd like to move this to a better system in the future but you don't need to do anything in this PR

File: code/modules/mob/living/carbon/human/ai/ai_equipment.dm [hunk 2]
```diff
@@ -16,10 +16,24 @@
...
+	. = ..()
+	if(new_human.ai_brain)
+		new_human.ai_brain.appraise_inventory()
```
Comment (Zonespace27) [inline code/modules/mob/living/carbon/human/ai/ai_equipment.dm]:
```suggestion
	new_human.ai_brain?.appraise_inventory()
```

File: code/modules/mob/living/carbon/human/ai/brain/ai_brain.dm [hunk 3]
```diff
@@ -92,9 +92,11 @@ GLOBAL_LIST_EMPTY(human_ai_brains)
 
 	if(!currently_busy && primary_weapon && current_target && !currently_firing && COOLDOWN_FINISHED(src, fire_overload_cooldown) && primary_weapon.has_ammunition())
 		currently_busy = TRUE
-		if(get_dist(tied_human, current_target) > gun_data.maximum_range)
+		var/target_futile = HAS_TRAIT(current_target, TRAIT_INCAPACITATED)
```
Comment (Zonespace27) [inline code/modules/mob/living/carbon/human/ai/brain/ai_brain.dm]: is_mob_incapacitated() would be better to use than this

File: code/modules/mob/living/carbon/human/ai/brain/ai_brain.dm [hunk 4]
```diff
@@ -92,9 +92,11 @@ GLOBAL_LIST_EMPTY(human_ai_brains)
...
+		if(get_dist(tied_human, current_target) > gun_data.optimal_range || target_futile)
 			if(!has_ongoing_action(/datum/ongoing_action/approach_target) && !in_cover)
-				ADD_ONGOING_ACTION(src, /datum/ongoing_action/approach_target, current_target, gun_data.maximum_range)
+				var/walk_distance = target_futile ? gun_data.minimum_range : gun_data.optimal_range
```
Comment (Zonespace27) [inline code/modules/mob/living/carbon/human/ai/brain/ai_brain.dm:98]: Will this cause them to walk into (possible) enemy fire to try and kill a downed member?

File: code/modules/mob/living/carbon/human/ai/brain/ai_brain_items.dm [hunk 5]
```diff
@@ -116,7 +116,7 @@
 
 	tried_reload = FALSE // We don't really need to do this in a smart way
 	if(belt)
-		if(!istype(tied_human.belt, /obj/item/storage/belt))
+		if(!istype(tied_human.belt, /obj/item/storage)) // belts can be backpacks, don't ask
```
Comment (Zonespace27) [inline code/modules/mob/living/carbon/human/ai/brain/ai_brain_items.dm:119]: :clueless:

File: code/modules/mob/living/carbon/human/ai/brain/ai_brain_targeting.dm [hunk 6]
```diff
@@ -140,32 +140,21 @@
...
-		shotgun.pump_shotgun(tied_human)
-		shotgun.recent_pump = world.time
-	else if(istype(primary_weapon, /obj/item/weapon/gun/boltaction) && !primary_weapon.in_chamber)
-		var/obj/item/weapon/gun/boltaction/bolt = primary_weapon
-		bolt.unique_action(tied_human)
-		bolt.recent_cycle = world.time
-		bolt.unique_action(tied_human)
-		bolt.recent_cycle = world.time
-	if(!primary_weapon.in_chamber || !friendly_check())
+	gun_data.before_fire(primary_weapon, tied_human, src)
```
Comment (Zonespace27) [inline code/modules/mob/living/carbon/human/ai/brain/ai_brain_targeting.dm:150]: Much better than the system I had

File: code/modules/mob/living/carbon/human/ai/firearm_appraisal.dm [hunk 7]
```diff
@@ -8,14 +8,24 @@ GLOBAL_LIST_INIT_TYPED(firearm_appraisals, /datum/firearm_appraisal, build_firea
...
 	/// How many rounds to fire in 1 burst at most
 	var/burst_amount_max = 8
 	/// List of types that set the human AI to this appraisal type
 	var/list/gun_types = list()
 
+/// List of things we do before our next fire based on weapon type
+/datum/firearm_appraisal/proc/before_fire(obj/item/weapon/gun/firearm, mob/living/carbon/user, datum/human_ai_brain/AI)
+	SHOULD_CALL_PARENT(TRUE) // Every weapon can be twohanded
+	AI.ensure_primary_hand(firearm)
+	if(firearm.flags_item & TWOHANDED && !(firearm.flags_item & WIELDED))
```
Comment (Zonespace27) [inline code/modules/mob/living/carbon/human/ai/firearm_appraisal.dm]:
```suggestion
	if((firearm.flags_item & TWOHANDED) && !(firearm.flags_item & WIELDED))
```

File: code/modules/projectiles/guns/flamer/flamer.dm [hunk 8]
```diff
@@ -143,6 +143,8 @@
 	else
 		user.track_shot(initial(name))
 		unleash_flame(target, user)
+		current_mag.current_rounds = current_mag.get_ammo_percent()
+		SEND_SIGNAL(user, COMSIG_MOB_FIRED_GUN, src) // required by human AI
```
Comment (Zonespace27) [inline code/modules/projectiles/guns/flamer/flamer.dm]:
```suggestion
		SEND_SIGNAL(user, COMSIG_MOB_FIRED_GUN, src)
```

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/human/ai/brain/ai_brain.dm (hunk 3 above)
Comment (Danilcus-W) [reply code/modules/mob/living/carbon/human/ai/brain/ai_brain.dm]: done

File: code/modules/mob/living/carbon/human/ai/brain/ai_brain.dm (hunk 4 above)
Comment (Danilcus-W) [reply code/modules/mob/living/carbon/human/ai/brain/ai_brain.dm:98]: Nah, when AI is being shot it resets it's current_target to the firer ref, so to try and stay away from new threat when we'll implement retreating on min_range

File: code/modules/mob/living/carbon/human/ai/brain/ai_brain_items.dm (hunk 5 above)
Comment (Danilcus-W) [reply code/modules/mob/living/carbon/human/ai/brain/ai_brain_items.dm:119]: :cat:

File: code/modules/mob/living/carbon/human/ai/brain/ai_brain_targeting.dm (hunk 6 above)
Comment (Danilcus-W) [reply code/modules/mob/living/carbon/human/ai/brain/ai_brain_targeting.dm:150]: :smiling_face_with_3_hearts:

**Review by Zonespace27 — APPROVED** [review]

**Pull request merged**